### PR TITLE
Progressively narrow file suffixes.

### LINKED ISSUE #313: Wrong extension from filename
refers to #174, given a file name of `rgba.u1.strict_disabled.avif`, the extension is found to be `.u1.strict_disabled.avif` but should be `.avif`.

```py
│ /home/prjemian/micromamba/envs/tiled/lib/python3.10/site-packages/tiled/adapters/files.py:692 in │
│ _reader_factory_for_file                                                                         │
│                                                                                                  │
│   689 │   # path, so it can consider the file name and even open the file. It is also            │
│   690 │   # passed the mimetype determined above, or None if no match was found.                 │
│   691 │   if mimetype_detection_hook is not None:                                                │
│ ❱ 692 │   │   mimetype = mimetype_detection_hook(path, mimetype)                                 │
│   693 │   if mimetype is None:                                                                   │
│   694 │   │   msg = (                                                                            │
│   695 │   │   │   f"The file at {path} has a file extension {ext} this is not "                  │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │                       _ = None                                                               │ │
│ │                     ext = '.u1.strict_disabled.avif'                                         │ │
│ │                mimetype = None                                                               │ │
│ │ mimetype_detection_hook = <function detect_mimetype at 0x7f8984a82e60>                       │ │
│ │   mimetypes_by_file_ext = ChainMap({'.dat': 'text/spec_data', '.h5': 'application/x-hdf5',   │ │
│ │                           '.hdf': 'application/x-hdf5', '.webp': 'image/webp', '.avif':      │ │
│ │                           'image/avif'}, {'.h5': 'application/x-hdf5', '.nxs':               │ │
│ │                           'application/x-hdf5', '.hdf': 'application/x-hdf5', '.hdf5':       │ │
│ │                           'application/x-hdf5', '.csv': 'text/csv'})                         │ │
│ │                    path = PosixPath('data/images/rgba.u1.strict_disabled.avif')              │ │
│ │     readers_by_mimetype = ChainMap({'image/avif': <function read_image at 0x7f89787337f0>,   │ │
│ │                           'image/bmp': <function read_image at 0x7f89787337f0>, 'image/gif': │ │
│ │                           <function read_image at 0x7f89787337f0>, 'image/jpeg': <function   │ │
│ │                           read_image at 0x7f89787337f0>, 'image/png': <function read_image   │ │
│ │                           at 0x7f89787337f0>, 'image/tiff': <function read_image at          │ │
│ │                           0x7f89787337f0>, 'image/vnd.microsoft.icon': <function read_image  │ │
│ │                           at 0x7f89787337f0>, 'text/spec_data': <function read_spec_data at  │ │
│ │                           0x7f897827ba30>, 'image/webp': <function read_image at             │ │
│ │                           0x7f89787337f0>}, <OneShotCachedMap({'image/tiff': <lazy>,         │ │
│ │                           'text/csv': <lazy>,                                                │ │
│ │                           'application/vnd.openxmlformats-officedocument.spreadsheetml.shee… │ │
│ │                           <lazy>, 'application/x-hdf5': <bound method HDF5Adapter.from_file  │ │
│ │                           of <class 'tiled.adapters.hdf5.HDF5Adapter'>>})>)                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /home/prjemian/Documents/projects/prjemian/tiled-demo/custom.py:33 in detect_mimetype            │
│                                                                                                  │
│   30 │   │   mimetype = "text/csv"  # the default                                                │
│   31 │   │   for tester, mtype in mimetype_table.items():                                        │
│   32 │   │   │   # iterate through our set of known types                                        │
│ ❱ 33 │   │   │   if tester(filename):                                                            │
│   34 │   │   │   │   mimetype = mtype                                                            │
│   35 │   │   │   │   break                                                                       │
│   36                                                                                             │
│                                                                                                  │
│ ╭───────────────────────────── locals ─────────────────────────────╮                             │
│ │ filename = PosixPath('data/images/rgba.u1.strict_disabled.avif') │                             │
│ │ mimetype = 'text/csv'                                            │                             │
│ │    mtype = 'text/dhtioc_data'                                    │                             │
│ │   tester = <function is_dhtioc_file at 0x7f8978733640>           │                             │
│ ╰──────────────────────────────────────────────────────────────────╯                             │
│                                                                                                  │
│ /home/prjemian/Documents/projects/prjemian/tiled-demo/dhtioc_csv.py:35 in is_dhtioc_file         │
│                                                                                                  │
│    32 │   │   1615010402.53 51.5 20.6                                                            │
│    33 │   """                                                                                    │
│    34 │   with open(filename, "r") as fp:                                                        │
│ ❱  35 │   │   for line_number, content in enumerate(fp.read().splitlines()):                     │
│    36 │   │   │   if "dhtioc" in content:                                                        │
│    37 │   │   │   │   return True                                                                │
│    38 │   │   │   if line_number > 3:                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ filename = PosixPath('data/images/rgba.u1.strict_disabled.avif')                             │ │
│ │       fp = <_io.TextIOWrapper name='data/images/rgba.u1.strict_disabled.avif' mode='r'       │ │
│ │            encoding='UTF-8'>                                                                 │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /home/prjemian/micromamba/envs/tiled/lib/python3.10/codecs.py:322 in decode                      │
│                                                                                                  │
│    319 │   def decode(self, input, final=False):                                                 │
│    320 │   │   # decode input (taking the buffer into account)                                   │
│    321 │   │   data = self.buffer + input                                                        │
│ ❱  322 │   │   (result, consumed) = self._buffer_decode(data, self.errors, final)                │
│    323 │   │   # keep undecoded input until the next call                                        │
│    324 │   │   self.buffer = data[consumed:]                                                     │
│    325 │   │   return result                                                                     │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │  data = b'\x00\x00\x00\x18ftypavif\x00\x00\x00\x00avifmif1\x00\x00\x01xmeta\x00\x00\x00\x00… │ │
│ │ final = True                                                                                 │ │
│ │ input = b'\x00\x00\x00\x18ftypavif\x00\x00\x00\x00avifmif1\x00\x00\x01xmeta\x00\x00\x00\x00… │ │
│ │  self = <encodings.utf_8.IncrementalDecoder object at 0x7f897826a170>                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x98 in position 106: invalid start byte
```

### PR DESCRIPTION
**Pull request by danielballan**
Closes #313

### CODE REVIEW

File: tiled/adapters/files.py [hunk 1]
```diff
@@ -674,11 +674,16 @@ def _reader_factory_for_file(
...
-    # User-specified mapping from file extension to mimetype
-    # gets priority.
-    if ext in mimetypes_by_file_ext:
-        mimetype = mimetypes_by_file_ext[ext]
+    # For compound suffixes like '.u1.strict_disabled.avif' (a real example)
+    # consider in order:
+    # '.u1.strict_disabled.avif'
+    # '.strict_disabled.avif'
+    # '.avif'
+    for i in range(len(path.suffixes) - 1):
```
Comment (prjemian) [inline tiled/adapters/files.py:682]: Remove the ` - 1`

```py
In [52]: path = pathlib.Path("rgba.u1.strict_disabled.avif")

In [53]: path.suffixes
Out[53]: ['.u1', '.strict_disabled', '.avif']

In [56]:     for i in range(len(path.suffixes) - 1):
    ...:         ext = "".join(path.suffixes[i:])  # e.g. ".h5" or ".tar.gz"
    ...:         print(f"{ext=}")
    ...: 
ext='.u1.strict_disabled.avif'
ext='.strict_disabled.avif'

In [57]:     for i in range(len(path.suffixes)):
    ...:         ext = "".join(path.suffixes[i:])  # e.g. ".h5" or ".tar.gz"
    ...:         print(f"{ext=}")
    ...: 
ext='.u1.strict_disabled.avif'
ext='.strict_disabled.avif'
ext='.avif'
```

With this change, the patch works for me.

Comment (danielballan) [reply tiled/adapters/files.py]: Oops, I had it right the first time. :-D 

Fixed now. Thanks.

**Comment by prjemian** [comment]
Remove the ` - 1` as noted above.

**Comment by danielballan** [comment]
Done! In the original PR I had _added_ the `-1` in a separate commit. I rebased to remove that commit from history (to avoid future confusion) and force-pushed.

**Pull request merged**